planner/cascades/implementation_rules.go

```diff
@@ -151,23 +155,72 @@ func (r *ImplTiKVSingleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *pr
 	return impl.NewTableReaderImpl(reader, sg.Source.TblColHists), nil
 }
 
+// ImplTiKVDoubleReadGather implements TiKVDoubleGather as PhysicalIndexLookUpReader.
+type ImplTiKVDoubleReadGather struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTiKVDoubleReadGather) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	if prop.IsEmpty() {
+		return true
+	}
+	reader := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	if all, _ := prop.AllSameOrder(); all {
+		for i, col := range reader.IndexCols {
+			if col.Equal(nil, prop.Items[0].Col) {
+				return plannercore.MatchIndicesProp(reader.IndexCols[i:], reader.IndexColLens[i:], prop.Items)
+			}
+		}
+	}
+	return false
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplTiKVDoubleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	logicProp := expr.Group.Prop
+	dg := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	var reader plannercore.PhysicalPlan
+	var proj *plannercore.PhysicalProjection
+	reader = dg.GetPhysicalIndexLookUpReader(logicProp.Schema, logicProp.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt), reqProp.Clone(), reqProp.Clone())
+	if dg.HandleCol.ID == model.ExtraHandleID && !reqProp.IsEmpty() {
+		// TODO: if the origin schema for IndexLookUpReader has the `HandleCol`, we can remove the duplicated append.
+		reader.Schema().Append(dg.HandleCol)
+		reader.(*plannercore.PhysicalIndexLookUpReader).ExtraHandleCol = dg.HandleCol
+		proj = plannercore.PhysicalProjection{Exprs: expression.Column2Exprs(logicProp.Schema.Columns)}.Init(dg.SCtx(), logicProp.Stats, dg.SelectBlockOffset(), reqProp.Clone(), reqProp.Clone())
+		proj.SetSchema(logicProp.Schema)
+		proj.SetChildren(reader)
+	}
+	indexLookUp := impl.NewIndexLookUpReaderImpl(reader, dg.Source.TblColHists, proj)
+	indexLookUp.KeepOrder = !reqProp.IsEmpty()
+	return indexLookUp, nil
+}
+
 // ImplTableScan implements TableScan as PhysicalTableScan.
 type ImplTableScan struct {
 }
 
 // Match implements ImplementationRule Match interface.
 func (r *ImplTableScan) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
 	ts := expr.ExprNode.(*plannercore.LogicalTableScan)
-	return prop.IsEmpty() || (len(prop.Items) == 1 && ts.Handle != nil && prop.Items[0].Col.Equal(nil, ts.Handle))
+	return prop.IsEmpty() || ts.IsDoubleRead || (len(prop.Items) == 1 && ts.Handle != nil && prop.Items[0].Col.Equal(nil, ts.Handle))
```

Comment:
The problem is that, if the table path has a partial Aggregation, it will reject this property.

Reply:
That's the additional task for `PushAggDownGather`. No need to consider it in this PR.

Reply:
If we correctly set `prop` for `TableScan`, how can we let `TableScan` known it should to keep order for double read and append a `tidb_rowid` in schema.